mage_ai/streaming/sources/mongodb.py

```diff
@@ -0,0 +1,72 @@
+from dataclasses import dataclass
+from typing import Any, Callable, Mapping, Optional
+
+from bson.timestamp import Timestamp
+from pymongo import MongoClient
+from pymongo.typings import _Pipeline
+
+from mage_ai.shared.config import BaseConfig
+from mage_ai.streaming.sources.base import BaseSource
+
+
+@dataclass
+class MongoDBConfig(BaseConfig):
+    connection_str: str
+    database: str
+    collection: str = None
+    batch_size: Optional[int] = 10
+    pipeline: Optional[_Pipeline] = None
+    operation_time: Optional[Timestamp] = None
+    start_after: Optional[Mapping[str, Any]] = None
+
+
+class MongoSource(BaseSource):
+    config_class = MongoDBConfig
+
+    def init_client(self):
+        self.client = MongoClient(self.config.connection_str)
+
+    def read(self, handler: Callable):
+        pass
+
+    def build_watch_args(self):
+        """Build the watch arguments based on config attributes."""
+        watch_args = {}
+
+        # Check each parameter and append to watch_args if present
+        if hasattr(self.config, 'batch_size') and self.config.batch_size is not None:
+            watch_args['batch_size'] = self.config.batch_size
+
+        if hasattr(self.config, 'pipeline') and self.config.pipeline is not None:
+            watch_args['pipeline'] = self.config.pipeline
+
+        if hasattr(self.config, 'operation_time') and self.config.operation_time is not None:
+            watch_args['operation_time'] = self.config.operation_time
+
+        if hasattr(self.config, 'start_after') and self.config.start_after is not None:
+            watch_args['start_after'] = self.config.start_after
+
+        return watch_args
+
+    def batch_read(self, handler: Callable):
+        self._print("Start getting message for MongoDB streaming.")
+        try:
+            db = self.client.get_database(self.config.database)
+            parsed_messages = []
+            self._print(db)
+
+            if self.config.collection:
+                watch_args = self.build_watch_args()
+                collection = db[self.config.collection]
+
+                with collection.watch(**watch_args) as stream:
+                    for change in stream:
+                        parsed_messages.append(change)
```

Comment:
The parsed_messages is a list that keeps get new change appended.
So when you are writing the parsed_messages to destination, you're writing the duplicate messages.

Reply:
stream is continuously working and parsed_messages is not append. I don't understand clearly this problem.

Reply:
How can i see data in mage for checking? I change the code and i see that all changes.

Reply:
I have two alternatives to solve this problem.

Firstly i can't use parsed_messages list and directly send handler.

```
    def batch_read(self, handler: Callable):
        self._print("Start getting message for MongoDB streaming.")
        try:
            db = self.client.get_database(self.config.database)
            self._print(db)

            if self.config.collection:
                watch_args = self.build_watch_args()
                collection = db[self.config.collection]

                with collection.watch(**watch_args) as stream:
                    for change in stream:
                        self._print(f'Received a new message: {change}.')
                        handler([change])  

        except Exception as e:
            self._print(f"Error reading from MongoDB: {e}")
```

Secondly i use parsed_messages and after than to handler i clear parsed_messages

```
def batch_read(self, handler: Callable):
    self._print("Start getting message for MongoDB streaming.")
    try:
        db = self.client.get_database(self.config.database)
        parsed_messages = []
        self._print(db)

        if self.config.collection:
            watch_args = self.build_watch_args()
            collection = db[self.config.collection]

            with collection.watch(**watch_args) as stream:
                for change in stream:
                    parsed_messages.append(change)
                    self._print(f'Received {len(parsed_messages)} message. '
                                f'Sample: {parsed_messages[0]}.')
                    handler(parsed_messages)
                    parsed_messages.clear() 

    except Exception as e:
        self._print(f"Error reading from MongoDB: {e}")

```

Which one is good solution?

Reply:
either way works. but question: how is the `batch_size` param used here?

Reply:
Batch size is default i take 10 as i see another stream connectors. But MongoDB is recommended 100 batch size. Maybe i change this variable.

Reply:
One message

Reply:
I changed the function the first function. Can you review function if you have free time @wangxiaoyou1993?